[Mouse Jump] - screenshot performance (#24607)

mikeclayton opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

* #24607

## Composite Screenshot

Reduces the time required for **Mouse Jump** to display the initial desktop thumbnail image - instead of copying the entire bounding desktop rectangle (pink in image below) with ```CopyFromScreen``` it now uses multiple calls to native ```StretchBlt``` to capture the individual active areas of screens 1-6 and composite them into a thumbnail image.

![image](https://user-images.githubusercontent.com/1193763/223550380-58b4d2bf-2fc2-4346-8eae-d5f345c19b87.png)

## Progressive Activation

If the form takes >250ms to capture screenshots it trips over into a "progressive" mode where it shows the form with solid rectangles instead of screenshots so that the user sees *something* happening, and then continues to capture the screenshots in the background and displays them when they're ready:

https://user-images.githubusercontent.com/1193763/224168931-4bf5a2ad-687d-45a3-a97c-85330194bf85.mp4

## Background Texture

I've also added a slightly-graduated blue background to make it "pop" a bit more than the previous black background. This wasn't on the original ticket so I can take it out again if you prefer.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [X] **Closes:** #24607
- [X] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [X] **Tests:** Added/updated and all pass (no new tests)
- [X] **Localization:** All end user facing strings can be localized (no new strings)
- [X] **Dev docs:** Added/updated (no new dev info)
- [X] **New binaries:** Added on the required places - no new binaries
- [X] **Documentation updated:** If checked, please file a pull request on [our docs repo] (no new documentation) (https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

See #24607 for full details

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

Manual tests of updated Mouse Jump feature with 2 screens at different dpi scale, and 6 screens in a jagged diagonal layout per image above. Preview display time is noticeably improved for the 6 screen layout and no slower for the 2 screen layout.

mikeclayton commented:
@crutkas @jaimecbernardo - this PR is ready for review I think (assuming the build passes).

I need to re-add some unit tests that broke during this PR, but any thoughts about the core functionality welcome...

jaimecbernardo commented:
Hi @mikeclayton ,
Thanks for opening this PR.
I've given it a test but it seems that it's failing to teleport the mouse pointer to the right screen in my case.
Displays are 3 horizontal 1920x1080 side by side, where the middle one is the main display.
I suspect this might be related to the left screen having negative coordinates.
![image](https://user-images.githubusercontent.com/26118718/224308793-45ce5177-0b51-4e73-a70a-9deb3577a773.png)

Basically what I'm observing on Mouse Jump is that if I try to jump to screen 3, it jumps to screen 1. If I try to jump to screen 1, it jumps to screen 2. If I try to jump to screen 2, it seems to jump to screen 1 at its right edge, which is weird.

mikeclayton commented:
@jaimecbernardo - ah, yeah at a guess that's a regression on the primary screen not being the top-left-most again.

I'll take a look later today and try to get my unit tests working again as they would have caught that if they were actually working :-S

mikeclayton commented:
@jaimecbernardo - I think I've fixed the issue with the primary monitor (I rearranged my screens and it worked fine this time), and I've fixed the broken unit tests as well, so it's ready for another look when you get a chance.

Cheers,

M

mikeclayton commented:
Btw, I've found a rounding issue which means the preview form fails to load on some combinations of screen layouts. I've got a fix which I'll push this evening, if you want to hold off reviewing until that's ready...

mikeclayton commented:
I've fixed the rounding issue - that's the last problem I know about so this is ready for review again now...

stefansjfw reviewed (approved):
LGTM! Everything works well!

There are conflicts here after merging this fix https://github.com/microsoft/PowerToys/pull/24527/files. Can you resolve the conflicts? it can be merged after that

mikeclayton commented:
> LGTM! Everything works well!
> 
> There are conflicts here after merging this fix https://github.com/microsoft/PowerToys/pull/24527/files. Can you resolve the conflicts? it can be merged after that

@stefansjfw - I suspected there would be :-).

I’ll merge from main into here and resolve the conflict tonight.

mikeclayton commented:
I've merged main into this PR and fixed the conflicts.

Currently waiting for the build to finish - if it works then it's good to go, or if it fails I'll take a look tomorrow...

stefansjfw reviewed (approved):
Looks good! Thanks for the contribution!
